pkg/storage/pkg/decomposedfs/metadata/hybrid_backend.go

```diff
@@ -0,0 +1,466 @@
+	"github.com/opencloud-eu/reva/v2/pkg/storage/utils/filelocks"
+)
+
+var _metadataOffloadedAttr = prefixes.OcPrefix + "metadata_offloaded"
+
+type MetadataPathFunc func(MetadataNode) string
+
+// HybridBackend stores the file attributes in extended attributes
+type HybridBackend struct {
+	offloadLimit     int
+	metaCache        cache.FileMetadataCache
+	metadataPathFunc MetadataPathFunc
+}
+
+// NewMessageBackend returns a new HybridBackend instance
+func NewHybridBackend(offloadLimit int, metadataPathFunc MetadataPathFunc, o cache.Config) HybridBackend {
+	return HybridBackend{
+		offloadLimit:     offloadLimit,
+		metaCache:        cache.GetFileMetadataCache(o),
+		metadataPathFunc: metadataPathFunc,
+	}
+}
+
+// Name returns the name of the backend
+func (HybridBackend) Name() string { return "hybrid" }
+
+// IdentifyPath returns the space id, node id and mtime of a file
+func (b HybridBackend) IdentifyPath(_ context.Context, path string) (string, string, time.Time, error) {
+	spaceID, _ := xattr.Get(path, prefixes.SpaceIDAttr)
+	id, _ := xattr.Get(path, prefixes.IDAttr)
+
+	mtimeAttr, _ := xattr.Get(path, prefixes.MTimeAttr)
+	mtime, _ := time.Parse(time.RFC3339Nano, string(mtimeAttr))
+	return string(spaceID), string(id), mtime, nil
+}
+
+// Get an extended attribute value for the given key
+// No file locking is involved here as reading a single xattr is
+// considered to be atomic.
+func (b HybridBackend) Get(ctx context.Context, n MetadataNode, key string) ([]byte, error) {
+	attribs := map[string][]byte{}
+	err := b.metaCache.PullFromCache(b.cacheKey(n), &attribs)
+	if err == nil && len(attribs[key]) > 0 {
+		return attribs[key], err
+	}
+
+	if isOffloadingAttribute(key) {
+		// check if grants are offloaded
+		offloaded, err := xattr.Get(n.InternalPath(), _metadataOffloadedAttr)
+		if err == nil && string(offloaded) == "1" {
+			msgpackAttribs := map[string][]byte{}
+			msgBytes, err := os.ReadFile(b.MetadataPath(n))
+			if err != nil {
+				return nil, err
+			}
+			err = msgpack.Unmarshal(msgBytes, &msgpackAttribs)
+			if err != nil {
+				return nil, err
+			}
+			return msgpackAttribs[key], nil
+		}
+	}
+	return xattr.Get(n.InternalPath(), key)
+}
+
+// GetInt64 reads a string as int64 from the xattrs
+func (b HybridBackend) GetInt64(ctx context.Context, n MetadataNode, key string) (int64, error) {
+	attr, err := b.Get(ctx, n, key)
+	if err != nil {
+		return 0, err
+	}
+	v, err := strconv.ParseInt(string(attr), 10, 64)
+	if err != nil {
+		return 0, err
+	}
+	return v, nil
+}
+
+// List retrieves a list of names of extended attributes associated with the
+// given path in the file system.
+func (b HybridBackend) List(ctx context.Context, n MetadataNode) (attribs []string, err error) {
+	return b.list(ctx, n, true)
+}
+
+func (b HybridBackend) list(ctx context.Context, n MetadataNode, acquireLock bool) (attribs []string, err error) {
+	filePath := n.InternalPath()
+	attrs, err := xattr.List(filePath)
+	if err == nil {
```

Comment:
shouldn't we also read the offloaded attributes

Reply:
well ... we can actually get rid of List() https://github.com/opencloud-eu/reva/pull/48